File: src/go/problems/max_profit_2_test.go
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type MaxProfit2 struct {}

func TestMaxProfit2(t *testing.T) {
	gen := core.TestSuite[MaxProfit2]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{1,3,2,8,4,9}).Param(2).Result(8)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{1,3,7,5,10,3}).Param(3).Result(6)
	}).Run(t)
}

func (MaxProfit2) Solution(prices []int, fee int) int {    
    max := func(left int, right int) int {
        if left > right {
            return left
        }
        return right
    }

    temp := -prices[0]
    profit := 0

    for i := 1; i < len(prices); i++ {
        
        temp = max(temp, profit - prices[i])
        profit = max(profit, temp + prices[i] - fee)
    }

    return profit
}

File: src/python/problems/max_profit2_test.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii

from core.problem_base import *

class MaxProfit2(ProblemBase):
    def Solution(self, prices: list[int]) -> int:
        profit = 0

        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                profit += prices[i] - prices[i - 1]

        return profit

if __name__ == '__main__':
    TestGen(MaxProfit2) \
        .Add(lambda tc: tc.Param([7,1,5,3,6,4]).Result(7)) \
        .Add(lambda tc: tc.Param([1,2,3,4,5]).Result(4)) \
        .Add(lambda tc: tc.Param([7,6,4,3,1]).Result(0)) \
        .Run()
